### ISSUE TITLE
Implementation into openpi

### ISSUE THREAD

**Issue by rohan-bansal** [Dexmal/realtime-vla#8]
Hey there, this is a great work!

What would it take to implement this into the openpi repository? I've currently created a new model file in `src/openpi/models` based off of `BaseModel`, except it calls `Pi0Infererence.forward` after processing observations in the `sample_actions()` function instead of proceeding with denoising steps. 

```python
    @override
    @torch.no_grad()
    def sample_actions(self, device, observation: _model.Observation, 
                    noise=None, num_steps=10) -> torch.Tensor:
        
        bsize = observation.state.shape[0]
        if noise is None:
            actions_shape = (bsize, self.config.action_horizon, self.config.action_dim)
            noise = self.sample_noise(actions_shape, device)

        # use pi0_pytorch preprocessing func
        images, img_masks, lang_tokens, lang_masks, state = self._preprocess_observation(observation, train=False)
        

        images = images[:self.num_views]
        images_converted = []
        for img in images:
            img = img[0]
            img = img.permute(1, 2, 0)  # Convert to (H, W, C)
            images_converted.append(img)
        
        images_stacked = torch.stack(images_converted, dim=0)  # (num_views, H, W, C)
        images_bf16 = images_stacked.to(dtype=torch.bfloat16, device='cuda')
        
        state_bf16 = state[0].to(dtype=torch.bfloat16, device='cuda')
        noise_input = noise[0].to(dtype=torch.bfloat16, device='cuda')
        
        # Run inference
        output_actions = self.inference_engine.forward(
            images_bf16,
            state_bf16,
            noise_input
        )
        
        # (1, action_horizon, action_dim)
        return output_actions.unsqueeze(0).to(dtype=torch.float32)
```

I'm noticing problems in simulation rollout (LIBERO), where robot motion is unintelligible. What am I doing wrong here?

**Comment by Dexmal** [comment]
Your code snippet looks fine to us. For further inspection, can you provide the following data files (by google drive?) so that we can work on this issue?
- Checkpoint generated by the openpi codebase
- The input tensors (images_bf16, state_bf16, noise_input)
- The desired output tensors (when you do not use our kernels)
We will try to reproduce the mismatch and see what might go wrong.

**Comment by mayunchao1994** [comment]
Okay, it's the official pi0_libero model. I mistakenly thought it was a model you finetuned yourself. Let me fix this issue.

**Comment by rohan-bansal** [comment]
@mayunchao1994 can you please provide your rollout code? I seem to still be encountering issues rolling out. Additionally, you unnormalized the images but according to the pi0_infer input, it requires a normalized image?

**Comment by mayunchao1994** [comment]
Based on your visualization, it looks like `extra_delta_transform` is causing a translation offset in `actions[..., :6]` between JAX and Triton. The gripper values are reasonable since they're not delta. In the official pi0 code, after sample_actions, it calls `outputs = self._output_transform(outputs)`, which includes `actions[..., :6] = actions[..., :6] + observation["state"][..., :6]`. I suspect delta_transform might be applied twice. Please also verify that `outputs = self._output_transform(outputs)` is consistent between the Triton and JAX versions. Additionally, you can try running the comparison results using the script I provided to see if it matches your script's results. You can also print the input to the model (output after input_transform), the model output, and the final output after output_transform to facilitate debugging.

**Comment by mayunchao1994** [comment]
I found a bug: `decoder_x_buf` should have shape `(chunk_size, 1024)`, not `(decoder_seq_len, 1024)`. I’ve updated the code. Please re-test the diff after aligning the input/output handling mentioned above.

**Comment by rohan-bansal** [comment]
Awesome thanks @mayunchao1994! After aligning the input/output transforms I was still getting weird results, but the `decoder_x_buf` issue was what fixed it for me. Here is the comparison graph now:

<img width="2084" height="3092" alt="Image" src="https://github.com/user-attachments/assets/7b354e8e-ffdc-4050-981b-98c64273c1c8" />

Is the flatlining between steps 200 and 250 for RY expected?

Additionally, now that this works, I intend to implement batching support.

**Comment by mayunchao1994** [comment]
Thanks a lot for raising this issue and providing a test case. It really helped me track down and fix the bug.
I think the differences at steps 200 and 250 for RY are expected, since the Triton version uses bf16 throughout, and some numerical precision differences vs. the JAX version are normal.
Implementing batching support would be very useful. We’re open to PRs — feel free to send a patch.
